src/mlpack/core/cv/metrics/roc_auc_score_impl.hpp

```diff
@@ -0,0 +1,78 @@
+/**
+ * @file core/cv/metrics/roc_auc_score_impl.hpp
+ * @author Sri Madhan M
+ *
+ * Implementation of the area under Receiver Operating Characteristic curve
+ * (ROC-AUC) score.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_CV_METRICS_ROCAUCSCORE_IMPL_HPP
+#define MLPACK_CORE_CV_METRICS_ROCAUCSCORE_IMPL_HPP
+
+#include <mlpack/core/cv/metrics/accuracy.hpp>
+#include <iostream>
+
+namespace mlpack {
+namespace cv {
+
+template<size_t PC /* PositiveClass */>
+template<typename MLAlgorithm, typename DataType>
+double ROC_AUC<PC>::Evaluate(MLAlgorithm& model,
+                              const DataType& data,
+                              const arma::Row<size_t>& labels)
+{
+  util::CheckSameSizes(data, labels, "ROC_AUC::Evaluate()");
+
+  arma::mat probabilities;
+  model.Classify(data, probabilities);
+
+  // compute labels with "1" for positive class and "0" for the other
+  arma::Col<size_t> binaryLabels = arma::conv_to<arma::Col<size_t>>::from(
+                                        ((arma::umat) (labels == PC)).row(0));
+
+  // probability scores of positive class
+  arma::Col<double> scoresOfPC   = arma::conv_to<arma::Col<double>>::from(
+                                        probabilities.row(PC));
+
+  size_t numberOfTrueLabels  = arma::sum(binaryLabels);
+  size_t numberOfFalseLabels = binaryLabels.n_rows - numberOfTrueLabels;
+
+  // sort labels and probabilities, using probability scores
+  arma::ucolvec sortedScoreIndices = arma::stable_sort_index(
+                                        scoresOfPC, "descend");
+  arma::Col<size_t> sortedLabels = binaryLabels(sortedScoreIndices);
+  arma::Col<double> sortedScores = scoresOfPC(sortedScoreIndices);
+
+  // compute indices of unique probability scores
+  arma::ucolvec uniqueScoreIndices = arma::find(arma::diff(sortedScores));
+  uniqueScoreIndices.insert_rows(uniqueScoreIndices.n_rows, 1);
+  uniqueScoreIndices(uniqueScoreIndices.n_rows - 1) = scoresOfPC.n_rows - 1;
+
+  // compute true positive rate, and false positive rate
+  arma::Col<size_t> cummulativeSum = arma::cumsum(sortedLabels);
+  cummulativeSum = cummulativeSum(uniqueScoreIndices);
+
+  arma::Col<double> tpr, fpr;
+  tpr = arma::conv_to<arma::Col<double>>::from(cummulativeSum);
+  fpr = 1 + uniqueScoreIndices - tpr;
+  tpr /= numberOfTrueLabels;
+  fpr /= numberOfFalseLabels;
+
+  // to ensure that the (fpr, tpr) starts at (0, 0)
+  tpr.insert_rows(0, 1);
+  fpr.insert_rows(0, 1);
+  tpr(0) = fpr(0) = 0;
+
+  // compute area under the curve using trapezoidal rule
+  arma::mat auc = arma::trapz(fpr, tpr);
```

Comment:
I'm not convinced the trapezoid rule is the best way to compute AUC here.  Can you justify the choice?  Maybe there is something I overlooked.

Reply:
Since we need to find the AUC for some set of points (which will be monotonically increasing), my choices were to compute it manually or using numerical integration. And **Armadillo** provides the `arma::trapz()`, so I went with the trapezoidal rule.

If there is something wrong with using the trapezoidal rule here, please point it out.

Reply:
Thanks for the explanation.  I think it is actually better to go with a rectangular integration approach here.  Let me explain why, and let me know if there are any issues with my reasoning.

The AUC is, literally, the 'area under the curve' of the ROC curve.  The ROC curve is defined by a series of points `(fpr, tpr)` containing true-positive rates vs. false-positive rates.  You are right that the TPR must only monotonically increase, as well as the FPR.  Let's take an extreme case: suppose we only have the two points `(0.0, 0.0)` and `(1.0, 1.0)`; that is, we know that we can either have 0% TPR and 0% FPR, or 100% TPR and 100% FPR.  So, the question is, what should the AUC be in this case?  We need to be pessimistic when we compute this, so if we were to, e.g., set our classifier such that it gives 50% FPR (halfway between our two points), we can't assume that we have 50% TPR---we only know that we have 0% TPR.  Thus, the AUC here should be 0, not the quantity we'd get using the trapezoid rule.

In the same vein, note that the Wikipedia page for 'receiver operating characteristic' (https://en.wikipedia.org/wiki/Receiver_operating_characteristic) displays three lines, all of which assume a constant TPR between any two FPR points.  Integrating these lines to produce the AUC would need rectangular integration, not the trapezoidal rule.

Anyway, hopefully what I wrote makes sense---let me know if I can clarify anything.  I think it should be easy to adapt the code.